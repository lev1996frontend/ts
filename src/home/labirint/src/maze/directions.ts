
export const mapDirections = ['top', 'right', 'bottom', 'left'] as const
export type MapDirection = typeof mapDirections[number]

export const playerDirections = [ 'left', 'forward', 'right', 'back'] as const
export type PlayerDirection = typeof playerDirections[number]

export const reverseDirection = (direction: MapDirection): MapDirection => {
	switch(direction) {
		case "top":
			return "bottom"
		case "bottom":
			return "top"
		case "left":
			return "right"
		case "right":
			return "left"
	}
}

export const directionMapToPlayer = (
	mapDirection: MapDirection,
	lookDirection: MapDirection
): PlayerDirection => {
	switch (lookDirection) {
		case "top": switch (mapDirection) {
			case "top": return 'forward'
			case "bottom": return 'back'
			case "left": return 'left'
			case "right": return 'right'
		}
		case "bottom": switch (mapDirection) {
			case "top": return 'back'
			case "bottom": return 'forward'
			case "left": return 'right'
			case "right": return 'left'
		}
		case "right": switch (mapDirection) {
			case "top": return 'left'
			case "bottom": return 'right'
			case "left": return 'back'
			case "right": return 'forward'
		}
		case "left": switch (mapDirection) {
			case "top": return 'right'
			case "bottom": return 'left'
			case "left": return 'forward'
			case "right": return 'back'
		}
	}
}

export const directionPlayerToMap = (
	playerDirection: PlayerDirection,
	lookDirection: MapDirection
): MapDirection => {
	switch (lookDirection) {
		case "top": switch (playerDirection) {
			case "forward": return 'top'
			case "back": return 'bottom'
			case "left": return 'left'
			case "right": return 'right'
		}
		case "bottom": switch (playerDirection) {
			case "forward": return 'bottom'
			case "back": return 'top'
			case "left": return 'right'
			case "right": return 'left'
		}
		case "left": switch (playerDirection) {
			case "forward": return 'left'
			case "back": return 'right'
			case "left": return 'bottom'
			case "right": return 'top'
		}
		case "right": switch (playerDirection) {
			case "forward": return 'right'
			case "back": return 'left'
			case "left": return 'top'
			case "right": return 'bottom'
		}
	}
}
