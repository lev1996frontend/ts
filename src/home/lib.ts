export type Header = string | (() => string)

export const menu = (
	...args: (
		[items: Record<string, () => void>] |
		[header: Header, items: Record<string, () => void>]
	)
): void => {
	if(args.length === 1) {
		_menu('', ...args)
	} else {
		_menu(...args)
	}
}

const _menu = (header: Header, items: Record<string, () => void>): void => {
	while (true) {
		console.clear()
		if (header) {
			print(
				typeof header === 'string'
					? header
					: header()
			)
		}
		const itemsString = Object.keys(items)
		for (let i = 0; i < itemsString.length; i++) {
			print(`${i + 1}. ${itemsString[i]}`)
		}
		print('0. выйти')

    const commandIndex = +input()

		if (commandIndex === 0) {
			return 
		}

		const key = itemsString[commandIndex - 1]
		const action = key ? items[key] : undefined

		if (action) {
			console.clear()
			action()
		}
	}
}

export function select(arr: unknown[]): number | undefined
export function select(header: string, arr: unknown[]): number | undefined
export function select<T extends object>(obj: T): keyof T | undefined
export function select<T extends object>(header: string, obj: T): keyof T | undefined
export function select<T extends object>(
	...args: (
		[arr: unknown[]] | [header: string, arr: unknown[]] 
		| [obj: T] | [header: string, obj: T]
	)
) {
	if (args.length === 1) {
		return _select('', args[0])
	}
	return _select(args[0], args[1])
}

const _select = <T extends object>(header: string, obj: T | unknown[]): string | number | undefined => {
	while (true) {
		console.clear()
		
		if (header) {
			print(header)
		}

		const isArray = Array.isArray(obj)
		const array = isArray ? obj : Object.keys(obj)
		if (!isArray) {
			print()
			array.forEach((key, index) => {
				print(`${index + 1}: ${key}`)
			})
			print()
		}
		const n = inputFromRange(0, array.length, {
			message: 'Выберите значение',
			hint: '0 - отмена',
		})
		if (n === 0) {
			return undefined
		}
		console.clear()
		return (
			isArray
				? n - 1
				: array[n - 1] as string
		)
	}
}

export const inputNumber = (header?: string, validation?: (item: number) => boolean | string): number => {
	while (true) {
		const inputStr = input(header)
		const inputNum = +inputStr

		const result = validation?.(inputNum) ?? true

		if (inputStr !== '' && result === true && isFinite(inputNum)) {
			return inputNum
		}
		
		print(
			typeof result === 'string' && result !== ''
				? result
				: 'Введите корректное число!'
		)
	}
}

type RangeOptions = {
	message: string
	showRange: boolean
	hint: string
}

export const inputFromRange = (
	min: number,
	max: number,
	options?: Readonly<Partial<RangeOptions>>
): number => {
	const message = options?.message ?? 'Введите число'
	const showRange = options?.showRange ?? true
	const hint = options?.hint ?? ''
	if (min < max) {
		const text = (
			message +
			`${showRange ? ` от ${min} до ${max}` : ''}`+
			`${hint ? ` (${hint})` : ''}`
		)
		return inputNumber(text, (value) => {
			if(value < min || value > max) {
				return `Error: число должно быть от ${min} до ${max}`
			}
			return true
		})
	}
	throw new Error(`Error: inputFromRange(${min}, ${max})`)
}
