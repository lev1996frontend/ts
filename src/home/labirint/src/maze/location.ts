import { MapDirection, mapDirections } from "./directions"

export type Obstacle = boolean | number // есть ли препятствие или дверь с номером ключа

export const isDoor = (obstacle: Obstacle) => typeof obstacle === 'number'

export type Location = Record<MapDirection, Obstacle> & {
	key?: number
}

export const availableDirections = (location: Location, keys: number[]): MapDirection[] => {
	const directions: MapDirection[] = []

	for (const mapDir of mapDirections) {
		if (!location[mapDir] || isDoor(location[mapDir]) && keys.includes(location[mapDir])) {
			directions.push(mapDir)
		}
	}

  return directions
}
