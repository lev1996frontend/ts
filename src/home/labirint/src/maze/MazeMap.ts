import { MapDirection } from "./directions"
import { Location } from "./location"

type ReadonlyMap = readonly (readonly Location[])[]

type Position = {
	row: number
	column: number
}

export type MazeEntries = {
	start: Position
	finish: Position
}

export class MazeMap {
	readonly start: Position
	readonly finish: Position

	readonly #map: ReadonlyMap

	constructor(map: ReadonlyMap, entries: MazeEntries) {
		if (!map[0]?.length) {
			throw new Error('empty map')
		}

		this.#map = map
		this.start = entries.start
		this.finish = entries.finish
	}

	cell(position: Position): Location | null {
		const row = position.row - 1
		const column = position.column - 1

		if (
			row < 0 || row >= this.#map.length ||
			column < 0 || column >= this.#map[0]!.length
		) {
			return null
		}

		return this.#map[row]![column]!

	}

	isStartCell(position: Position) {
		return  position.row === this.start.row && position.column === this.start.column
	}

	isFinishCell(position: Position) {
		return  position.row === this.finish.row && position.column === this.finish.column
	}

  // не хватает проверок кординаты
  getStartLookDirection(): MapDirection {
    if (this.start.column === 0) {
      return 'right'
    }

    if (this.start.row === 0) {
      return 'bottom'
    }

    if (this.start.column > this.start.row) {
      return 'left'
    }

    return 'top'
  }
}
