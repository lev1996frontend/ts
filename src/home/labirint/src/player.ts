import { MapDirection } from "./maze/directions"

export type Player = {
	readonly name: string // может не понадобится
	readonly keys: number[]
	readonly doorsOpened: number[]
	row: number
	column: number
	lookDirection: MapDirection
}
