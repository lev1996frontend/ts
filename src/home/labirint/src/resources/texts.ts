import {
  directionMapToPlayer,
  directionPlayerToMap,
  MapDirection,
  mapDirections,
  PlayerDirection,
  playerDirections
} from '../maze/directions'

import {
  isDoor,
  Location
} from '../maze/location'

import { Player } from '../player'

// TODO: ?
export const translatedMoveDirection = ['прямо', 'назад', 'налево', 'направо'] as const
export type TranslatedMoveDirection = typeof translatedMoveDirection[number]
export const translatedDirections = {
	forward: 'прямо',
	back: 'назад',
	left: 'налево',
	right: 'направо',
} satisfies Record<PlayerDirection, TranslatedMoveDirection>
export const translateDirections = (directions: PlayerDirection[]) => {
	return directions.map(dir => translatedDirections[dir])
}
export const parseDirection = (direction: TranslatedMoveDirection): PlayerDirection => {
	for (const key in translatedDirections) {
		const playerDirection = key as PlayerDirection
		if (translatedDirections[playerDirection] === direction) {
			return playerDirection
		}
	}
	throw new Error('Таких направлений нет')
}

// Move Player

const exit = ' и выходишь из лабиринта'

const doorPassed = (doorIsOpen: boolean): string => {
  return doorIsOpen
    ? ' (проходишь через открытую дверь)'
    : ' (открываешь ключом и проходишь закрытую дверь)'
}

const deadEnd = ' и оказываешься в тупике.'
const fork = ' и выходишь на развилку.'

const keyFound = '\nПод ногами ты находишь ключ.'
const keyFoundStop = ' и под ногами находишь ключ.'

const translatedNextDirections = {
  forward: 'перед тобой',
  left: 'слева',
  right: 'справа',
} satisfies Record<Exclude<PlayerDirection, 'back'>, string>

const passedByClosedDoor = (currentCell: Location, lookDirection: MapDirection): string => {
  let description = ` мимо закрытой двери`

  for (const playerDirection of playerDirections) {
    const direction = directionPlayerToMap(playerDirection, lookDirection)
    if (playerDirection !== 'back' && isDoor(currentCell[direction])) {
      description += ` ${translatedNextDirections[playerDirection]} `
    }
  }

  return description
}

const closedDoorsFound = (currentCell: Location, player: Player): string => {
  let hasDoors = false
  let description = ' '

  for (const direction of mapDirections) {
    if (isDoor(currentCell[direction]) && !player.keys.includes(currentCell[direction])) {
      const doorDirection = directionMapToPlayer(direction, player.lookDirection)
      const translatedDoorDirection = translatedDirections[doorDirection]
      description += (
        translatedDoorDirection.charAt(0).toUpperCase() + translatedDoorDirection.slice(1) +
        ' от тебя' +
        (hasDoors ? ` ещё одна` : '') +
        ` закрытая дверь, от которой у тебя нет ключа.`
      )
      hasDoors = true
    }
  }
  return description
}

const stepsCase = (stepsForward: number): string => {
	const rigthDigit = stepsForward % 10
	if (rigthDigit === 1) {
		return 'шаг'
	} else if (rigthDigit >= 2 && rigthDigit <= 4) {
		return 'шага'
	} else if (rigthDigit >= 11 && rigthDigit <= 14) {
		return 'шагов'
	}	else {
		return 'шагов'
	}
}

const forwardSteps = (): string  => {
  return ''
}

const forwardStepsText = (steps: number): string  => {
  let description = ', идешь '
  if (steps > 1) {
    description += `${steps} ${stepsCase(steps)} `
  }
  description += 'прямо'

  return description
}

const turnStep = (direction: 'left' | 'right', repeat: boolean): string => {
  let description = ', '
  if (repeat) {
    description += 'затем снова '
  }
  description += (
    `поворачиваешь ` +
    (direction === "left" ? "налево" : "направо")
  )

  return description
}

const moveBack = ' разворачиваешься назад'

const withStartText = (text: string) => {
  return 'Ты' + text.replace(/^,/, '')
}

export const createDescription = () => {
  let forwardStepsCount = 0

  return new Proxy({
    exit,
    doorPassed,
    deadEnd,
    fork,
    keyFound,
    keyFoundStop,
    passedByClosedDoor,
    closedDoorsFound,
    forwardSteps,
    turnStep,
    moveBack,
    withStartText,
  } satisfies Record<string, string | ((...args: any[]) => string)>, {
    get(target, key) {
      let description = ''

      if (key === forwardSteps.name) {
        forwardStepsCount++
      } else if (forwardStepsCount) {
        description += forwardStepsText(forwardStepsCount)
        forwardStepsCount = 0
      }

      const text = (target as any)[key]

      if (typeof text === 'function') {
        return (...args: any[]) => {
          return description + text(...args)
        }
      }
      return description + text
    }
  })
}
