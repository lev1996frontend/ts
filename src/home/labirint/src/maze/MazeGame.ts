import { Player } from '../player'
import { directionMapToPlayer, directionPlayerToMap, mapDirections, PlayerDirection, playerDirections, reverseDirection } from './directions'
import { availableDirections, isDoor, Location, Obstacle } from './location'
import { MazeMap } from './MazeMap'
import { createDescription } from '../resources/texts'

export class OutError extends Error {
	constructor() {
		super('out of maze')
	}
}

export class MazeGame {
  #map: MazeMap
  #player: Player

	constructor(map: MazeMap, player: Player) {
		this.#map = map
		this.#player = player
	}

	get inMaze(): boolean {
		return !!this.#map.cell(this.#player)
	}

	get atFinish(): boolean {
		return (
			this.#player.row === this.#map.finish.row
			&& this.#player.column === this.#map.finish.column
		)
	}

	get atStart(): boolean {
		return (
			this.#player.row === this.#map.start.row
			&& this.#player.column === this.#map.start.column
		)
	}

	getAvailableMoveDirections(): PlayerDirection[] {
		const currentCell = this.#map.cell(this.#player)

		if (!currentCell) {
			throw new OutError()
		}

		const lookDirection = this.#player.lookDirection
		const availableDirections: PlayerDirection[] = []

    // ?
		const canPass = (obstacle: Obstacle) => (
			!obstacle
			|| isDoor(obstacle) && this.#player.keys.includes(obstacle)
		)

    for (const direction of playerDirections) {
      const mapDirection = directionPlayerToMap(direction, lookDirection)
      const obstacle = currentCell[mapDirection]
      if (canPass(obstacle)) {
        availableDirections.push(direction)
      }
    }

		return availableDirections
	}

	// TODO: перемещение игрока
  movePlayer(moveDirection: PlayerDirection): string {
    const texts = createDescription()

    let description = ''

    let prevDirection: PlayerDirection | undefined

    let nextDirection = directionPlayerToMap(moveDirection, this.#player.lookDirection)

    while (true) {
      switch (nextDirection) {
        case 'left':
          this.#player.column--
          break
        case 'right':
          this.#player.column++
          break
        case 'top':
          this.#player.row--
          break
        case 'bottom':
          this.#player.row++
          break
      }

      const passedPlayerDirection = directionMapToPlayer(nextDirection, this.#player.lookDirection)

      if (passedPlayerDirection === 'left' || passedPlayerDirection === 'right') {
        description += texts.turnStep(
          passedPlayerDirection,
          prevDirection === passedPlayerDirection
        )
      }
      else if (passedPlayerDirection === 'forward') {
        description += texts.forwardSteps()
      }
      else {
        description += texts.moveBack
      }

      const currentLocation = this.#map.cell(this.#player)

      if (!currentLocation) {
        description += texts.exit
        break
      }

      const backDirection = reverseDirection(nextDirection)

      if (isDoor(currentLocation[backDirection])) {
        const door = currentLocation[backDirection]
        const doorIsOpen = this.#player.doorsOpened.includes(door)

        if (!doorIsOpen) {
          this.#player.doorsOpened.push(door)
        }

        description += texts.doorPassed(doorIsOpen)
      }

      const currentDirections = availableDirections(currentLocation, this.#player.keys)

      if (currentDirections.length === 1) {
        description += texts.deadEnd
        if (this.takeKeyIfExists(currentLocation)) {
          description += texts.keyFound
        }
        description += texts.closedDoorsFound(currentLocation, this.#player)
        break
      }

      if (currentDirections.length > 2) {
        description += texts.fork
        if (this.takeKeyIfExists(currentLocation)) {
          description += texts.keyFound
        }
        description += texts.closedDoorsFound(currentLocation, this.#player)
        break
      }

			if (this.takeKeyIfExists(currentLocation)) {
        description += texts.keyFoundStop
        description += texts.closedDoorsFound(currentLocation, this.#player)
				break
			}

      if (hasClosedDoor(currentLocation, this.#player)) {
        description += texts.passedByClosedDoor(currentLocation, this.#player.lookDirection)
      }

      this.#player.lookDirection = nextDirection

      nextDirection = currentDirections.filter(d => d !== backDirection)[0]!
    }

    this.#player.lookDirection = nextDirection

		return texts.withStartText(description)
	}

	takeKeyIfExists(currentLocation: Location): boolean { // boolean
		if (typeof currentLocation.key === 'number') {
			this.#player.keys.push(currentLocation.key)
			delete currentLocation.key
			return true
		}
		return false
	}

	// TODO: remove
	locationDescription(): string {
		// ...
		return (
			`row = ${this.#player.row}, `+
			`column = ${this.#player.column}, `+
			`look direction = ${this.#player.lookDirection}, `+
			`directions = ${this.getAvailableMoveDirections()}, ` +
			`keys = ${this.#player.keys || 0}`
		)
	}
}

//

const isDoorClosed = (player: Player, door: number) => {
  return !player.keys.includes(door)
}
const hasClosedDoor = (cell: Location, player: Player) => {
  return mapDirections.some(direction =>
    isDoor(cell[direction])
    && isDoorClosed(player, cell[direction])
  )
}
