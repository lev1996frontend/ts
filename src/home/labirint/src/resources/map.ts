import { MazeMap } from "../maze/MazeMap";


export const map = new MazeMap([
	[ // 1.1
		{	top: true, bottom: false, left: true, right: false },
		// 1.2
		{	top: true, bottom: true, left: false, right: false },
		// 1.3
		{	top: true, bottom: false, left: false, right: true },
		// 1.4
		{	top: true, bottom: false, left: true, right: false },
		// 1.5
		{	top: true, bottom: false, left: false, right: false },
		// 1.6
		{	top: true, bottom: true, left: false, right: false },
		// 1.7
		{	top: true, bottom: true, left: false, right: false },
		// 1.8
		{	top: true, bottom: true, left: false, right: true },
		// 1.9
		{	top: true, bottom: false, left: true, right: false },
		// 1.10
		{	top: true, bottom: false, left: false, right: true },
	],
	// 2
	[ // 2.1
		{	top: false, bottom: true, left: true, right: false },
		// 2.2
		{	top: true, bottom: true, left: false, right: true },
		// 2.3
		{	top: false, bottom: false, left: true, right: true },
		// 2.4
		{	top: false, bottom: false, left: true, right: true },
		// 2.5
		{	top: false, bottom: true, left: true, right: false },
		// 2.6
		{	top: true, bottom: true, left: false, right: false },
		// 2.7
		{	top: true, bottom: true, left: false, right: false },
		// 2.8
		{	top: true, bottom: false, left: false, right: true },
		// 2.9
		{	top: false, bottom: false, left: true, right: true },
		// 2.10
		{	top: false, bottom: false, left: true, right: true },
	],
	// 3
	[ // 3.1
		{	top: true, bottom: false, left: true, right: true, key: 1},
		// 3.2
		{	top: true, bottom: false, left: true, right: false },
		// 3.3
		{	top: false, bottom: true, left: false, right: true },
		// 3.4
		{	top: false, bottom: true, left: true, right: false },
		// 3.5
		{	top: true, bottom: true, left: false, right: false },
		// 3.6
		{	top: true, bottom: false, left: false, right: true },
		// 3.7
		{	top: true, bottom: false, left: true, right: true },
		// 3.8
		{	top: false, bottom: false, left: true, right: true },
		// 3.9
		{	top: false, bottom: false, left: true, right: true },
		// 3.10
		{	top: false, bottom: false, left: true, right: true },
	],
	// 4
	[ // 4.1
		{	top: false, bottom: false, left: true, right: true },
		// 4.2
		{	top: false, bottom: true, left: true, right: false },
		// 4.3
		{	top: true, bottom: true, left: false, right: false },
		// 4.4
		{	top: true, bottom: false, left: false, right: true },
		// 4.5
		{	top: true, bottom: false, left: true, right: false },
		// 4.6
		{	top: false, bottom: true, left: false, right: false },
		// 4.7
		{	top: false, bottom: false, left: false, right: true },
		// 4.8
		{	top: false, bottom: false, left: true, right: false },
		// 4.9
		{	top: false, bottom: true, left: false, right: true },
		// 4.10
		{	top: false, bottom: false, left: true, right: true },
	],
	// 5
	[ // 5.1
		{	top: false, bottom: true, left: true, right: false },
		// 5.2
		{	top: true, bottom: false, left: false, right: false },
		// 5.3
		{	top: false, bottom: false, left: true, right: false },
		// 5.4
		{	top: false, bottom: true, left: true, right: false },
		// 5.5
		{	top: false, bottom: true, left: false, right: true },
		// 5.6
		{	top: true, bottom: false, left: true, right: false },
		// 5.7
		{	top: false, bottom: true, left: false, right: true },
		// 5.8
		{	top: false, bottom: true, left: true, right: false },
		// 5.9
		{	top: true, bottom: true, left: false, right: true },
		// 5.10
		{	top: false, bottom: false, left: true, right: true },
	],
	// 6
	[ // 6.1
		{	top: true, bottom: false, left: true, right: true },
		// 6.2
		{	top: false, bottom: true, left: true, right: false },
		// 6.3
		{	top: true, bottom: true, left: false, right: false },
		// 6.4
		{	top: true, bottom: true, left: false, right: false },
		// 6.5
		{	top: true, bottom: false, left: false, right: false },
		// 6.6
		{	top: false, bottom: true, left: false, right: true },
		// 6.7
		{	top: true, bottom: false, left: true, right: true },
		// 6.8
		{	top: true, bottom: false, left: true, right: false },
		// 6.9
		{	top: true, bottom: false, left: false, right: true },
		// 6.10
		{	top: false, bottom: false, left: true, right: 2 },
	],
	// 7
	[ // 7.1
		{	top: false, bottom: true, left: true, right: false },
		// 7.2
		{	top: true, bottom: true, left: false, right: false },
		// 7.3
		{	top: true, bottom: false, left: false, right: false },
		// 7.4
		{	top: true, bottom: true, left: false, right: true },
		// 7.5
		{	top: false, bottom: false, left: true, right: true },
		// 7.6
		{	top: true, bottom: false, left: true, right: false },
		// 7.7
		{	top: false, bottom: true, left: false, right: false },
		// 7.8
		{	top: false, bottom: false, left: false, right: true },
		// 7.9
		{	top: false, bottom: true, left: true, right: false },
		// 7.10
		{	top: false, bottom: true, left: false, right: true },
	],
	// 8
	[ // 8.1
		{	top: true, bottom: false, left: true, right: false },
		// 8.2
		{	top: true, bottom: true, left: false, right: false },
		// 8.3
		{	top: false, bottom: true, left: false, right: false },
		// 8.4
		{	top: true, bottom: true, left: false, right: false },
		// 8.5
		{	top: false, bottom: true, left: false, right: true },
		// 8.6
		{	top: false, bottom: false, left: true, right: false },
		// 8.7
		{	top: true, bottom: false, left: false, right: true },
		// 8.8
		{	top: false, bottom: false, left: true, right: true },
		// 8.9
		{	top: true, bottom: true, left: true, right: false },
		// 8.10
		{	top: true, bottom: false, left: false, right: true },
	],
	// 9
	[ // 9.1
		{	top: false, bottom: true, left: true, right: true },
		// 9.2
		{	top: true, bottom: false, left: true, right: false },
		// 9.3
		{	top: true, bottom: true, left: false, right: false },
		// 9.4
		{	top: true, bottom: false, left: false, right: false },
		// 9.5
		{	top: true, bottom: false, left: false, right: true },
		// 9.6
		{	top: false, bottom: false, left: true, right: true },
		// 9.7
		{	top: false, bottom: false, left: true, right: true },
		// 9.8
		{	top: false, bottom: true, left: true, right: false },
		// 9.9
		{	top: true, bottom: true, left: false, right: false },
		// 9.10
		{	top: false, bottom: true, left: false, right: true },
	],
	// 10
	[ // 10.1
		{	top: true, bottom: true, left: false, right: false },
		// 10.2
		{	top: false, bottom: true, left: false, right: true },
		// 10.3
		{	top: true, bottom: true, left: true, right: false },
		// 10.4
		{	top: false, bottom: true, left: false, right: true },
		// 10.5
		{	top: false, bottom: true, left: true, right: false },
		// 10.6
		{	top: false, bottom: true, left: false, right: true },
		// 10.7
		{	top: false, bottom: true, left: true, right: false },
		// 10.8
		{	top: true, bottom: true, left: false, right: false },
		// 10.9
		{	top: true, bottom: true, left: false, right: 1 },
		// 10.10
		{	top: true, bottom: true, left: 1, right: true, key: 2	},
	],
], {
	start: { row: 10, column: 0 },
	finish: {	row: 6, column: 11 },
})