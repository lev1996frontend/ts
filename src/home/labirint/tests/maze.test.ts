import { map } from '../src/resources/map'
import { MazeGame } from './../src/maze/MazeGame'
import { Player } from '../src/player'

// MazeGame Player map

describe('лабиринт - перемещение игрока', () => {
  test('на стартовую локацию', () => {

    const player = createPlayer()
    const maze = new MazeGame(map, player)
    const movingDescription = maze.movePlayer('forward')

    expect(movingDescription).toMatchSnapshot()
    expect(player.row).toBe(9)
    expect(player.column).toBe(4)
    // TODO: getAvailableMoveDirections
  })

  // TODO: Ты поворачиваешь направо, "затем снова" поворачиваешь направо...
  test.skip('к закрытой двери без ключа', () => {
    const player = createPlayer({
      row: 8,
      column: 6,
      lookDirection: 'top',
    })
    const maze = new MazeGame(map, player)
    const movingDescription = maze.movePlayer('right')

    expect(movingDescription).toMatchSnapshot()
    expect(player.row).toBe(10)
    expect(player.column).toBe(9)
    // TODO: getAvailableMoveDirections
  })

  test.skip('к закрытой двери с ключом', () => {
    const player = createPlayer({
      row: 8,
      column: 6,
      lookDirection: 'top',
      keys: [1],
    })
    const maze = new MazeGame(map, player)
    const movingDescription = maze.movePlayer('right')

    expect(movingDescription).toMatchSnapshot()
    expect(player.row).toBe(10)
    expect(player.column).toBe(10)
    // TODO: getAvailableMoveDirections

    expect(player.doorsOpened).toContain(1)
    expect(player.keys).toContain(2)
  })

  // TODO: выход из лабиринта на старте (проигрыш)
  // TODO: выход из лабиринта в конце (выигрыш)
})

function createPlayer(player: Partial<Player> = {}): Player {
  return {
    name: 'playerName',
    row: map.start.row,
    column: map.start.column,
    lookDirection: map.getStartLookDirection(),
    keys: [],
    doorsOpened: [],
    ...player,
  }
}

const n1 = [1,2,3]
const n2 = [4,5,6]
const n3 = [...n1, ...n2]
