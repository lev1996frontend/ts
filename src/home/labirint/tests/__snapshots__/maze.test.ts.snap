// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`лабиринт - перемещение игрока к закрытой двери без ключа 1`] = `"Ты поворачиваешь направо, затем снова поворачиваешь направо, идешь прямо, поворачиваешь налево, идешь прямо и оказываешься в тупике. Прямо от тебя закрытая дверь, от которой у тебя нет ключа."`;

exports[`лабиринт - перемещение игрока к закрытой двери с ключом 1`] = `
"Ты поворачиваешь направо, затем снова поворачиваешь направо, идешь прямо, поворачиваешь налево, идешь 2 шага прямо (открываешь ключом и проходишь закрытую дверь) и оказываешься в тупике.
Под ногами ты находишь ключ."
`;

exports[`лабиринт - перемещение игрока на стартовую локацию 1`] = `"Ты идешь 2 шага прямо, поворачиваешь налево, поворачиваешь направо, идешь прямо и выходишь на развилку."`;
